dotnet/src/SemanticKernel/SkillDefinition/SKFunction.cs

```diff
@@ -231,24 +231,20 @@ public async Task<SKContext> InvokeAsync(SKContext context, CompleteRequestSetti
         {
             this.AddDefaultValues(context.Variables);
 
-            var resultContext = await this._function(this._aiService?.Value, settings ?? this._aiRequestSettings, context, cancellationToken).ConfigureAwait(false);
-            context.Variables.Update(resultContext.Variables);
+            return await this._function(this._aiService?.Value, settings ?? this._aiRequestSettings, context, cancellationToken).ConfigureAwait(false);
         }
-        else
+
+        try
         {
-            try
-            {
-                context = await this._function(null, settings, context, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception e) when (!e.IsCriticalException())
-            {
-                const string Message = "Something went wrong while executing the native function. Function: {0}. Error: {1}";
-                this._logger.LogError(e, Message, this._function.Method.Name, e.Message);
-                context.Fail(e.Message, e);
-            }
+            return await this._function(null, settings, context, cancellationToken).ConfigureAwait(false);
+        }
+        catch (Exception e) when (!e.IsCriticalException())
+        {
+            const string Message = "Something went wrong while executing the native function. Function: {0}. Error: {1}";
+            this._logger.LogError(e, Message, this._function.Method.Name, e.Message);
+            context.LastException = e;
```

Comment:
Same here, it looks like in some places we re-throw and in some places we just set `LastException` property. Not sure it's the scope of this PR though :)

Reply:
Yeah, I ran into this too. My first instinct was to just throw here, but that broke some of the UTs because they're expecting to successfully complete the "InvokeAsync" call and then check for failure in the context. I decided to keep the same functionality for now, and to overhaul it when we introduce the SKResult type. I also want to coordinate with @SergeyMenshykh about how such errors should propagate, and not overlap with his PR.